### PR TITLE
More ClickHouse logic

### PR DESCRIPTION
**Pull request by qoega**
1) Renamed Clickhouse->ClickHouse, clickhouse->clickHouse.
2) Returned `clickhouse-jdbc`
3) Tried to implement TypedExpressionGenerator instead of Untyped.
4) Implemented only Int32/String constants for now.
5) Use clickhouse-jdbc ClickHouseDataType instead of separate enum.
And more...

### CODE REVIEW

File: src/sqlancer/clickhouse/ClickHouseSchema.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        super(databaseTables);
+    }
+
+    public ClickHouseTables getRandomTableNonEmptyTables() {
+        return new ClickHouseTables(Randomly.nonEmptySubset(getDatabaseTables()));
+    }
+
+    private static ClickHouseLancerDataType getColumnType(String typeString) {
+        ClickHouseLancerDataType primitiveType;
+//        if (typeString.startsWith("Nullable(")) {
```
Comment (mrigger) [inline src/sqlancer/clickhouse/ClickHouseSchema.java:255]: If possible, avoid checking in commented-out code. We do not enforce this currently, since SQLancer's code contains already many instances, but I plan to remove all commented-out code at a later stage.

File: src/sqlancer/clickhouse/ClickHouseToStringVisitor.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    @Override
+    public void visit(ClickHouseSelect select, boolean inner) {
+        if (inner) {
+            sb.append("(");
+        }
+        sb.append("SELECT ");
+        visit(select.getFetchColumns());
+        sb.append(" FROM ");
+        visit(select.getFromList());
+        // if (!select.getFromList().isEmpty() /* && !select.getJoinExpressions().isEmpty() */) {
```
Comment (mrigger) [inline src/sqlancer/clickhouse/ClickHouseToStringVisitor.java:66]: You might also want to consider removing this commented-out code.

File: src/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperation.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            @Override
+            ClickHouseConstant apply(ClickHouseConstant left, ClickHouseConstant right) {
+                return left.applyLess(right);
+            }
+
+        },
+        SMALLER_EQUALS("<=") {
+            @Override
+            ClickHouseConstant apply(ClickHouseConstant left, ClickHouseConstant right) {
+                ClickHouseConstant lessThan = left.applyLess(right);
```
Comment (mrigger) [inline src/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperation.java:69]: I noticed that you started implementing all these `execute`/`apply` methods. Do you plan to also support PQS in a later stage? For NoREC and TLP, it is not necessary to implement these.

File: src/sqlancer/clickhouse/gen/ClickHouseCommon.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import sqlancer.clickhouse.ClickHouseProvider;
+import sqlancer.clickhouse.ClickHouseSchema;
+import sqlancer.clickhouse.ClickHouseVisitor;
+import sqlancer.clickhouse.ast.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public final class ClickHouseCommon {
```
Comment (mrigger) [inline src/sqlancer/clickhouse/gen/ClickHouseCommon.java:15]: It seems that a number of methods in this class are specific to SQLite. Can we delete those that we do not need for ClickHouse?

File: src/sqlancer/clickhouse/gen/ClickHouseCommon.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        if (Randomly.getBoolean()) {
+            sb.append("INDEXED BY ");
+            sb.append(indexName);
+        } else {
+            sb.append("NOT INDEXED");
+        }
+        return sb.toString();
+    }
+
+    public static String getFreeTableName(ClickHouseSchema s) {
```
Comment (mrigger) [inline src/sqlancer/clickhouse/gen/ClickHouseCommon.java:92]: The `getFreeTableName()` and similar methods are now available in https://github.com/sqlancer/sqlancer/blob/master/src/sqlancer/schema/AbstractSchema.java#L63. They should also be available in `ClickhouseSchema`, since it inherits from `AbstractSchema`. Could you please use those instead?

File: src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import sqlancer.clickhouse.gen.ClickHouseCommon;
+import sqlancer.clickhouse.gen.ClickHouseExpressionGenerator;
+
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ClickHouseTLPHavingOracle implements TestOracle {
```
Comment (mrigger) [inline src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java:21]: It is not necessary for this PR, but in the future you might want to consider inheriting from `ClickHouseTLPBase` and use the existing functionality there. I already implemented this for some of the other DBMS.

File: src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+        ClickHouseExpression predicate = gen.getHavingClause();
+        select.setHavingClause(predicate);
+        String firstQueryString = ClickHouseVisitor.asString(select);
+        select.setHavingClause(new ClickHouseUnaryPrefixOperation(predicate, ClickHouseUnaryPrefixOperation.ClickHouseUnaryPrefixOperator.NOT));
+        String secondQueryString = ClickHouseVisitor.asString(select);
+        select.setHavingClause(new ClickHouseUnaryPostfixOperation(predicate, ClickHouseUnaryPostfixOperation.ClickHouseUnaryPostfixOperator.IS_NULL, false));
+        String thirdQueryString = ClickHouseVisitor.asString(select);
+        String combinedString = firstQueryString + " UNION ALL " + secondQueryString + " UNION ALL " + thirdQueryString;
+        if (combinedString.contains("EXIST")) {
```
Comment (mrigger) [inline src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java:65]: I think this test for `EXIST` is not necessary. It was some workaround for `SQLite`, which I should remove asap.

File: src/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperation.java (hunk 3 above)
Comment (qoega) [reply src/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperation.java:69]: If I'll have enough time.

File: src/sqlancer/clickhouse/gen/ClickHouseCommon.java (hunk 5 above)
Comment (qoega) [reply src/sqlancer/clickhouse/gen/ClickHouseCommon.java]: Done

File: src/sqlancer/clickhouse/gen/ClickHouseCommon.java (hunk 4 above)
Comment (qoega) [reply src/sqlancer/clickhouse/gen/ClickHouseCommon.java:9]: Sure

File: src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java (hunk 7 above)
Comment (qoega) [reply src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java]: done

File: test/sqlancer/dbms/TestSQLite3.java [hunk 8]
```diff
@@ -12,7 +12,7 @@
     public void testSqlite() {
         // run with one thread due to multithreading issues, see https://github.com/sqlancer/sqlancer/pull/45
         assertEquals(0, Main.executeMain(new String[] { "--random-seed", "0", "--timeout-seconds", TestConfig.SECONDS,
-                "--num-threads", "1", "--num-queries", "0", "sqlite3" }));
+                "--num-threads", "1", "--num-queries", "100", "sqlite3" }));
```
Comment (mrigger) [inline test/sqlancer/dbms/TestSQLite3.java:15]: Could you please set `--num-queries` back to `0`? NoREC can still find bugs in the SQLite version included in the latest JDBC driver, so this test might fail sometimes.

File: test/sqlancer/dbms/TestConfig.java [hunk 9]
```diff
@@ -1,7 +1,6 @@
 package sqlancer.dbms;
 
 public class TestConfig {
-
     public static final String NUM_QUERIES = "1000";
-    public static final String SECONDS = "300";
+    public static final String SECONDS = "30";
```
Comment (mrigger) [inline test/sqlancer/dbms/TestConfig.java:5]: Can you set this back to `300`? `30` seconds is a bit short.

File: test/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperationTest.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+import org.junit.jupiter.api.Test;
+import ru.yandex.clickhouse.domain.ClickHouseDataType;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class ClickHouseBinaryComparisonOperationTest {
```
Comment (mrigger) [inline test/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperationTest.java:11]: Great, thanks for including these tests!

File: pom.xml [hunk 11]
```diff
@@ -125,22 +125,16 @@
...
-						<configuration>
-							<failurePriority>2</failurePriority>
-							<failOnViolation>true</failOnViolation>
-							<rulesets>
-								<ruleset>configs/pmd-rules.xml</ruleset>
-							</rulesets>
-						</configuration>
-					</execution>
-				</executions>
+				<goals>
```
Comment (mrigger) [inline pom.xml:128]: I just checked, and found that this change disabled the PMD checks. Could you please undo this change and fix the PMD warnings?

Comment (qoega) [reply pom.xml:128]: Actually IDEA showed errors about <rulesets> section. I'll revert it for now.

File: test/sqlancer/dbms/TestSQLite3.java (hunk 8 above)
Comment (qoega) [reply test/sqlancer/dbms/TestSQLite3.java:15]: Accidentally added changes that I used during debug. Reverted it.

File: test/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperationTest.java (hunk 10 above)
Comment (qoega) [reply test/sqlancer/clickhouse/ast/ClickHouseBinaryComparisonOperationTest.java:11]: That is almost nothing. Used it to debug my bugs

File: pom.xml (hunk 11 above)
Comment (mrigger) [reply pom.xml]: Okay, I see. This also ensures that the Travis CI gate only passes when no PMD warnings are issued.

**Comment by qoega** [comment]
Finally green Travis CI check!
